src/soroban/contract_client.ts

```diff
@@ -42,14 +37,61 @@ export type ContractClientOptions = {
    * })
    * ```
    */
-  wallet: Wallet;
+  wallet?: Wallet;
+
+  /**
+   * You can pass in `wallet` OR `account`, but not both. If you only pass
+   * `wallet`, `account` will be derived from it. If you can bypass this
+   * behavior by passing in your own account object.
+   */
+  account?: Account | Promise<Account>;
+};
+
+async function getPublicKey(wallet?: Wallet): Promise<string | undefined> {
+  if (!wallet) return undefined;
+  if (!(await wallet.isConnected()) || !(await wallet.isAllowed())) {
+    return undefined;
+  }
+  return (await wallet.getUserInfo()).publicKey;
+};
+
+export const NULL_ACCOUNT =
+  "GAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAWHF";
+
+/**
+ * Get account details from the Soroban network for the publicKey currently
+ * selected in user's wallet. If user is not connected to their wallet, {@link
+ * getPublicKey} returns undefined, and this will return {@link NULL_ACCOUNT}.
+ * This works for simulations, which is all that's needed for most view calls.
+ * If you want the transaction to be included in the ledger, you will need to
+ * provide a connected wallet.
+ */
+export async function getAccount(server: SorobanRpc.Server, wallet?: Wallet): Promise<Account> {
+  const publicKey = await getPublicKey(wallet);
+  return publicKey
+    ? await server.getAccount(publicKey)
+    : new Account(NULL_ACCOUNT, "0");
+};
+
+export type ContractClientOptions = AcceptsWalletOrAccount & {
+  contractId: string;
+  networkPassphrase: string;
+  rpcUrl: string;
+  errorTypes?: Record<number, { message: string }>;
 };
 
 export class ContractClient {
+  private server: SorobanRpc.Server;
+
   constructor(
     public readonly spec: ContractSpec,
     public readonly options: ContractClientOptions,
-  ) {}
+  ) {
+    this.server = new SorobanRpc.Server(this.options.rpcUrl, {
+      allowHttp: this.options.rpcUrl.startsWith("http://"),
```

Comment:
should we move on to forcing people to intentionally opt-in to insecure servers instead of doing it like this?